{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f594e55d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import shutil                            \n",
    "import glob                             \n",
    "import numpy as np\n",
    "import cv2\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import sklearn\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9e12b356",
   "metadata": {},
   "outputs": [],
   "source": [
    "from RNASeq_prep_n_RFOptimization import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e10d95d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80\n"
     ]
    }
   ],
   "source": [
    "os.getcwd()                            \n",
    "root_path = \"/home/io3247/GenFeatSel\"\n",
    "met_path = root_path + \"/genomic_blca/genomic_blca_met\"\n",
    "met_content = next(os.walk(met_path))[1]\n",
    "try:\n",
    "    met_content.remove(\".ipynb_checkpoints\")  \n",
    "except:\n",
    "    pass\n",
    "#print(met_content)\n",
    "print(len(met_content))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "73e6d17a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80\n"
     ]
    }
   ],
   "source": [
    "# Create a list containing path to all \"metastasis\" rna_seq files.\n",
    "met_list = []\n",
    "for barc_folder in met_content:\n",
    "    barc_folder_path = met_path + \"/\" + barc_folder\n",
    "    barc_folder_content = next(os.walk(barc_folder_path))[1]\n",
    "    met_folder = barc_folder_content[0]\n",
    "    met_folder_path = barc_folder_path + \"/\" + met_folder\n",
    "    met_extension = \"/*.tsv\"\n",
    "    rna_seq_files = glob.glob(met_folder_path + met_extension)     \n",
    "    met_list.append(rna_seq_files[0])\n",
    "    \n",
    "#print(met_list)\n",
    "print(len(met_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5684971d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4057c68d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80\n"
     ]
    }
   ],
   "source": [
    "no_met_path = root_path + \"/genomic_blca/genomic_blca_no-met\"\n",
    "no_met_content = next(os.walk(no_met_path))[1]  \n",
    "try:\n",
    "    no_met_content.remove(\".ipynb_checkpoints\")  \n",
    "except:\n",
    "    pass\n",
    "#print(no_met_content)\n",
    "print(len(no_met_content))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "135ec333",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80\n"
     ]
    }
   ],
   "source": [
    "# Create a list containing path to all \"no-metastasis\" rna_seq files.\n",
    "no_met_list = []\n",
    "for barc_folder in no_met_content:\n",
    "    barc_folder_path = no_met_path + \"/\" + barc_folder\n",
    "    barc_folder_content = next(os.walk(barc_folder_path))[1]\n",
    "    no_met_folder = barc_folder_content[0]\n",
    "    no_met_folder_path = barc_folder_path + \"/\" + no_met_folder\n",
    "    no_met_extension = \"/*.tsv\"\n",
    "    rna_seq_files = glob.glob(no_met_folder_path + no_met_extension)\n",
    "    no_met_list.append(rna_seq_files[0])\n",
    "    \n",
    "#print(no_met_list)\n",
    "print(len(no_met_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db9624eb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2282001e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 0 NaN values in dataframe.\n",
      "Samples values converted to float32\n",
      "There are 0 NaN values in dataframe after log transformation\n",
      "There are 463629 Inf values in data after log transformation\n",
      "0\n",
      "NaN and Inf values replaced with 0.\n",
      "Log transformation complete.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/.autofs/tools/spack/var/spack/environments/tumor-22092001/.spack-env/view/lib/python3.8/site-packages/pandas/core/internals/blocks.py:402: RuntimeWarning: divide by zero encountered in log10\n",
      "  result = func(self.values, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes with 0 value in greter than 80% of samples removed.\n",
      "Dataframe sorted by variance. Top 10,000 genes selected\n",
      "Dataframe transposed: Genes are now arranged in columns, and samples in rows\n",
      "Would you like to save transposed, preprocessed full fpkm dataframe? (y/n)n\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(160, 10000)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Multiple categories data pre-processing pipeline\n",
    "# Combine tpm_objs of all categories of samples cosidered into a sigle list before further pre-processing\n",
    "met_tpm_objs = all_tpm_objs_list(met_list, metastasis=True)\n",
    "met_objs = met_tpm_objs[0]\n",
    "met_labels = met_tpm_objs[1]\n",
    "no_met_tpm_objs = all_tpm_objs_list(no_met_list)\n",
    "no_met_objs = no_met_tpm_objs[0]\n",
    "no_met_labels = no_met_tpm_objs[1]\n",
    "all_cat_tpm_objs = met_objs + no_met_objs\n",
    "\n",
    "# Merge each rna-seq df into one full df\n",
    "all_cat_full_df = merge_dataframe(all_cat_tpm_objs)\n",
    "all_cat_full_df\n",
    "\n",
    "# Carry out further pre-processing- float32 type conversion, 0 values filter, variance filter, and transposition on df, log transformation\n",
    "prep_df = further_preprocessing(all_cat_full_df, save_filename_dot_csv = \"preprocessed_BLCA_tpm.csv\", apply_var_filter = True)\n",
    "prep_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "47eb0fcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>KRT6A</th>\n",
       "      <th>SPINK1</th>\n",
       "      <th>KRT5</th>\n",
       "      <th>KRT20</th>\n",
       "      <th>GSTM1</th>\n",
       "      <th>RPS4Y1</th>\n",
       "      <th>HMGCS2</th>\n",
       "      <th>KRT13</th>\n",
       "      <th>DSG3</th>\n",
       "      <th>KRT14</th>\n",
       "      <th>...</th>\n",
       "      <th>SEMA6C</th>\n",
       "      <th>TRIM14</th>\n",
       "      <th>RNF19B</th>\n",
       "      <th>G6PD</th>\n",
       "      <th>SPART</th>\n",
       "      <th>CAT</th>\n",
       "      <th>TMEM65</th>\n",
       "      <th>TNFRSF10A</th>\n",
       "      <th>TMTC2</th>\n",
       "      <th>metastasis</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>met_patient1</th>\n",
       "      <td>2.485847</td>\n",
       "      <td>0.468406</td>\n",
       "      <td>3.304415</td>\n",
       "      <td>2.314423</td>\n",
       "      <td>2.134161</td>\n",
       "      <td>2.771527</td>\n",
       "      <td>1.186009</td>\n",
       "      <td>3.853465</td>\n",
       "      <td>1.490822</td>\n",
       "      <td>1.525784</td>\n",
       "      <td>...</td>\n",
       "      <td>1.059692</td>\n",
       "      <td>1.523922</td>\n",
       "      <td>1.762159</td>\n",
       "      <td>1.778139</td>\n",
       "      <td>1.049319</td>\n",
       "      <td>1.775893</td>\n",
       "      <td>0.994396</td>\n",
       "      <td>1.488546</td>\n",
       "      <td>1.190326</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient2</th>\n",
       "      <td>3.770939</td>\n",
       "      <td>-0.588212</td>\n",
       "      <td>3.826497</td>\n",
       "      <td>-0.611189</td>\n",
       "      <td>0.199920</td>\n",
       "      <td>-1.266001</td>\n",
       "      <td>-0.863597</td>\n",
       "      <td>0.574922</td>\n",
       "      <td>1.556667</td>\n",
       "      <td>3.864502</td>\n",
       "      <td>...</td>\n",
       "      <td>0.627601</td>\n",
       "      <td>1.249193</td>\n",
       "      <td>1.821425</td>\n",
       "      <td>1.486739</td>\n",
       "      <td>1.187772</td>\n",
       "      <td>2.167517</td>\n",
       "      <td>1.162418</td>\n",
       "      <td>1.298386</td>\n",
       "      <td>0.722782</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient3</th>\n",
       "      <td>3.994303</td>\n",
       "      <td>0.708030</td>\n",
       "      <td>4.078915</td>\n",
       "      <td>0.324077</td>\n",
       "      <td>0.371012</td>\n",
       "      <td>2.030510</td>\n",
       "      <td>0.087568</td>\n",
       "      <td>1.213881</td>\n",
       "      <td>2.546487</td>\n",
       "      <td>4.447619</td>\n",
       "      <td>...</td>\n",
       "      <td>0.606231</td>\n",
       "      <td>1.454688</td>\n",
       "      <td>1.798896</td>\n",
       "      <td>2.091919</td>\n",
       "      <td>0.926867</td>\n",
       "      <td>1.780687</td>\n",
       "      <td>1.267814</td>\n",
       "      <td>1.325008</td>\n",
       "      <td>0.331529</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient4</th>\n",
       "      <td>-0.325139</td>\n",
       "      <td>2.241668</td>\n",
       "      <td>2.447783</td>\n",
       "      <td>1.007253</td>\n",
       "      <td>-1.037157</td>\n",
       "      <td>2.448398</td>\n",
       "      <td>2.709300</td>\n",
       "      <td>3.093846</td>\n",
       "      <td>0.222066</td>\n",
       "      <td>-0.467246</td>\n",
       "      <td>...</td>\n",
       "      <td>0.349627</td>\n",
       "      <td>1.470653</td>\n",
       "      <td>1.794031</td>\n",
       "      <td>2.121171</td>\n",
       "      <td>1.503799</td>\n",
       "      <td>2.153338</td>\n",
       "      <td>0.531185</td>\n",
       "      <td>0.917736</td>\n",
       "      <td>0.722206</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient5</th>\n",
       "      <td>3.101205</td>\n",
       "      <td>-0.693575</td>\n",
       "      <td>3.444901</td>\n",
       "      <td>-0.966174</td>\n",
       "      <td>1.443889</td>\n",
       "      <td>-0.173731</td>\n",
       "      <td>0.415991</td>\n",
       "      <td>1.454945</td>\n",
       "      <td>1.997646</td>\n",
       "      <td>3.041708</td>\n",
       "      <td>...</td>\n",
       "      <td>0.865187</td>\n",
       "      <td>1.732806</td>\n",
       "      <td>2.035776</td>\n",
       "      <td>1.774489</td>\n",
       "      <td>1.292398</td>\n",
       "      <td>1.829720</td>\n",
       "      <td>1.389975</td>\n",
       "      <td>0.898215</td>\n",
       "      <td>0.634024</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient6</th>\n",
       "      <td>2.855980</td>\n",
       "      <td>0.787723</td>\n",
       "      <td>3.565637</td>\n",
       "      <td>0.719679</td>\n",
       "      <td>2.169149</td>\n",
       "      <td>2.559360</td>\n",
       "      <td>0.508664</td>\n",
       "      <td>4.163777</td>\n",
       "      <td>2.718144</td>\n",
       "      <td>3.064817</td>\n",
       "      <td>...</td>\n",
       "      <td>0.912817</td>\n",
       "      <td>1.562614</td>\n",
       "      <td>1.905865</td>\n",
       "      <td>1.897888</td>\n",
       "      <td>1.176277</td>\n",
       "      <td>1.564493</td>\n",
       "      <td>1.159456</td>\n",
       "      <td>1.589704</td>\n",
       "      <td>1.295030</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient7</th>\n",
       "      <td>-0.035269</td>\n",
       "      <td>2.674077</td>\n",
       "      <td>1.923787</td>\n",
       "      <td>3.127244</td>\n",
       "      <td>-1.235824</td>\n",
       "      <td>2.743984</td>\n",
       "      <td>2.487469</td>\n",
       "      <td>2.634731</td>\n",
       "      <td>-1.406714</td>\n",
       "      <td>1.325243</td>\n",
       "      <td>...</td>\n",
       "      <td>0.487873</td>\n",
       "      <td>1.488315</td>\n",
       "      <td>1.648729</td>\n",
       "      <td>1.913074</td>\n",
       "      <td>1.545804</td>\n",
       "      <td>2.150972</td>\n",
       "      <td>0.809176</td>\n",
       "      <td>1.157566</td>\n",
       "      <td>1.140637</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient8</th>\n",
       "      <td>3.153689</td>\n",
       "      <td>1.715438</td>\n",
       "      <td>3.386103</td>\n",
       "      <td>2.503668</td>\n",
       "      <td>2.654869</td>\n",
       "      <td>-0.749824</td>\n",
       "      <td>0.985624</td>\n",
       "      <td>3.207443</td>\n",
       "      <td>1.572401</td>\n",
       "      <td>2.394732</td>\n",
       "      <td>...</td>\n",
       "      <td>1.181726</td>\n",
       "      <td>1.607846</td>\n",
       "      <td>1.720087</td>\n",
       "      <td>1.964063</td>\n",
       "      <td>1.360141</td>\n",
       "      <td>1.793545</td>\n",
       "      <td>1.140722</td>\n",
       "      <td>1.396809</td>\n",
       "      <td>1.027810</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient9</th>\n",
       "      <td>-1.369572</td>\n",
       "      <td>3.800762</td>\n",
       "      <td>-0.177309</td>\n",
       "      <td>3.007730</td>\n",
       "      <td>2.531929</td>\n",
       "      <td>2.315384</td>\n",
       "      <td>1.230280</td>\n",
       "      <td>3.240573</td>\n",
       "      <td>-1.950782</td>\n",
       "      <td>1.116309</td>\n",
       "      <td>...</td>\n",
       "      <td>0.708106</td>\n",
       "      <td>1.004085</td>\n",
       "      <td>1.750707</td>\n",
       "      <td>1.518511</td>\n",
       "      <td>1.423754</td>\n",
       "      <td>1.696444</td>\n",
       "      <td>0.753154</td>\n",
       "      <td>0.733927</td>\n",
       "      <td>1.140524</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient10</th>\n",
       "      <td>1.997679</td>\n",
       "      <td>3.144291</td>\n",
       "      <td>2.932373</td>\n",
       "      <td>0.647715</td>\n",
       "      <td>2.305203</td>\n",
       "      <td>-1.472370</td>\n",
       "      <td>2.299851</td>\n",
       "      <td>3.578766</td>\n",
       "      <td>0.059185</td>\n",
       "      <td>0.714925</td>\n",
       "      <td>...</td>\n",
       "      <td>0.558036</td>\n",
       "      <td>1.388144</td>\n",
       "      <td>1.522396</td>\n",
       "      <td>1.687928</td>\n",
       "      <td>1.363710</td>\n",
       "      <td>2.130172</td>\n",
       "      <td>0.939829</td>\n",
       "      <td>1.181572</td>\n",
       "      <td>1.117921</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient11</th>\n",
       "      <td>1.209936</td>\n",
       "      <td>3.451827</td>\n",
       "      <td>1.375446</td>\n",
       "      <td>2.356983</td>\n",
       "      <td>-1.411168</td>\n",
       "      <td>2.643656</td>\n",
       "      <td>2.336838</td>\n",
       "      <td>0.295193</td>\n",
       "      <td>-0.226360</td>\n",
       "      <td>1.575554</td>\n",
       "      <td>...</td>\n",
       "      <td>0.711723</td>\n",
       "      <td>0.953112</td>\n",
       "      <td>1.381830</td>\n",
       "      <td>1.603746</td>\n",
       "      <td>1.226966</td>\n",
       "      <td>1.742630</td>\n",
       "      <td>0.627438</td>\n",
       "      <td>0.698353</td>\n",
       "      <td>1.039676</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient12</th>\n",
       "      <td>3.481066</td>\n",
       "      <td>-0.289544</td>\n",
       "      <td>2.793193</td>\n",
       "      <td>0.248022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.183177</td>\n",
       "      <td>-0.833570</td>\n",
       "      <td>1.359475</td>\n",
       "      <td>1.699674</td>\n",
       "      <td>3.082666</td>\n",
       "      <td>...</td>\n",
       "      <td>0.925812</td>\n",
       "      <td>1.394826</td>\n",
       "      <td>1.848327</td>\n",
       "      <td>1.734852</td>\n",
       "      <td>1.507281</td>\n",
       "      <td>1.594602</td>\n",
       "      <td>1.027366</td>\n",
       "      <td>0.794174</td>\n",
       "      <td>0.867019</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient13</th>\n",
       "      <td>3.719666</td>\n",
       "      <td>1.002157</td>\n",
       "      <td>3.703100</td>\n",
       "      <td>-0.288530</td>\n",
       "      <td>-1.686133</td>\n",
       "      <td>2.459486</td>\n",
       "      <td>1.813181</td>\n",
       "      <td>3.418172</td>\n",
       "      <td>2.660139</td>\n",
       "      <td>2.928010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.154424</td>\n",
       "      <td>1.684059</td>\n",
       "      <td>2.136211</td>\n",
       "      <td>1.744812</td>\n",
       "      <td>1.484694</td>\n",
       "      <td>2.245799</td>\n",
       "      <td>1.018368</td>\n",
       "      <td>1.547062</td>\n",
       "      <td>0.655667</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient14</th>\n",
       "      <td>1.028050</td>\n",
       "      <td>0.479186</td>\n",
       "      <td>2.752916</td>\n",
       "      <td>0.240250</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.530136</td>\n",
       "      <td>-0.108965</td>\n",
       "      <td>3.454592</td>\n",
       "      <td>-0.680478</td>\n",
       "      <td>-0.018272</td>\n",
       "      <td>...</td>\n",
       "      <td>0.858814</td>\n",
       "      <td>1.345005</td>\n",
       "      <td>1.412870</td>\n",
       "      <td>1.832265</td>\n",
       "      <td>1.540573</td>\n",
       "      <td>1.591973</td>\n",
       "      <td>1.055818</td>\n",
       "      <td>0.898572</td>\n",
       "      <td>1.056024</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient15</th>\n",
       "      <td>1.308003</td>\n",
       "      <td>0.215532</td>\n",
       "      <td>0.778021</td>\n",
       "      <td>0.443857</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.461929</td>\n",
       "      <td>-0.934794</td>\n",
       "      <td>0.454555</td>\n",
       "      <td>-0.187020</td>\n",
       "      <td>2.129858</td>\n",
       "      <td>...</td>\n",
       "      <td>0.374565</td>\n",
       "      <td>1.567450</td>\n",
       "      <td>1.921197</td>\n",
       "      <td>1.875732</td>\n",
       "      <td>1.510821</td>\n",
       "      <td>1.487206</td>\n",
       "      <td>1.022486</td>\n",
       "      <td>1.020046</td>\n",
       "      <td>0.553325</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient16</th>\n",
       "      <td>2.717411</td>\n",
       "      <td>1.712626</td>\n",
       "      <td>3.099088</td>\n",
       "      <td>1.483767</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.303103</td>\n",
       "      <td>-0.348334</td>\n",
       "      <td>1.127898</td>\n",
       "      <td>1.216688</td>\n",
       "      <td>3.046589</td>\n",
       "      <td>...</td>\n",
       "      <td>0.745535</td>\n",
       "      <td>1.808487</td>\n",
       "      <td>1.798279</td>\n",
       "      <td>2.201339</td>\n",
       "      <td>1.533009</td>\n",
       "      <td>1.533213</td>\n",
       "      <td>0.952279</td>\n",
       "      <td>1.465845</td>\n",
       "      <td>0.472917</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient17</th>\n",
       "      <td>-0.963371</td>\n",
       "      <td>1.496707</td>\n",
       "      <td>0.348947</td>\n",
       "      <td>2.091970</td>\n",
       "      <td>-1.294992</td>\n",
       "      <td>2.585932</td>\n",
       "      <td>2.474985</td>\n",
       "      <td>3.059174</td>\n",
       "      <td>-1.642065</td>\n",
       "      <td>1.369774</td>\n",
       "      <td>...</td>\n",
       "      <td>1.007291</td>\n",
       "      <td>1.504158</td>\n",
       "      <td>1.926353</td>\n",
       "      <td>1.530493</td>\n",
       "      <td>1.439683</td>\n",
       "      <td>1.947224</td>\n",
       "      <td>1.007167</td>\n",
       "      <td>0.813094</td>\n",
       "      <td>1.593850</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient18</th>\n",
       "      <td>1.741497</td>\n",
       "      <td>1.882959</td>\n",
       "      <td>2.416238</td>\n",
       "      <td>2.547161</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.174769</td>\n",
       "      <td>2.836458</td>\n",
       "      <td>3.425118</td>\n",
       "      <td>0.167288</td>\n",
       "      <td>1.778171</td>\n",
       "      <td>...</td>\n",
       "      <td>0.586272</td>\n",
       "      <td>1.058915</td>\n",
       "      <td>1.118536</td>\n",
       "      <td>1.689522</td>\n",
       "      <td>0.779510</td>\n",
       "      <td>1.662170</td>\n",
       "      <td>0.004579</td>\n",
       "      <td>0.831678</td>\n",
       "      <td>0.634356</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient19</th>\n",
       "      <td>2.776374</td>\n",
       "      <td>-0.675923</td>\n",
       "      <td>2.472819</td>\n",
       "      <td>-0.726767</td>\n",
       "      <td>-0.367948</td>\n",
       "      <td>-0.876148</td>\n",
       "      <td>-0.377061</td>\n",
       "      <td>-0.131297</td>\n",
       "      <td>-0.142426</td>\n",
       "      <td>2.276137</td>\n",
       "      <td>...</td>\n",
       "      <td>0.882120</td>\n",
       "      <td>1.330328</td>\n",
       "      <td>1.785747</td>\n",
       "      <td>2.204198</td>\n",
       "      <td>0.375206</td>\n",
       "      <td>1.411414</td>\n",
       "      <td>0.711875</td>\n",
       "      <td>0.668265</td>\n",
       "      <td>0.459408</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>met_patient20</th>\n",
       "      <td>-1.563837</td>\n",
       "      <td>1.556350</td>\n",
       "      <td>-1.714443</td>\n",
       "      <td>2.449821</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.832605</td>\n",
       "      <td>1.841840</td>\n",
       "      <td>1.723930</td>\n",
       "      <td>-1.844664</td>\n",
       "      <td>-0.108686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.982416</td>\n",
       "      <td>1.693635</td>\n",
       "      <td>2.099919</td>\n",
       "      <td>1.756516</td>\n",
       "      <td>1.673222</td>\n",
       "      <td>2.366270</td>\n",
       "      <td>0.552778</td>\n",
       "      <td>0.973437</td>\n",
       "      <td>1.328308</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20 rows Ã— 10001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name         KRT6A    SPINK1      KRT5     KRT20     GSTM1    RPS4Y1  \\\n",
       "0                                                                           \n",
       "met_patient1   2.485847  0.468406  3.304415  2.314423  2.134161  2.771527   \n",
       "met_patient2   3.770939 -0.588212  3.826497 -0.611189  0.199920 -1.266001   \n",
       "met_patient3   3.994303  0.708030  4.078915  0.324077  0.371012  2.030510   \n",
       "met_patient4  -0.325139  2.241668  2.447783  1.007253 -1.037157  2.448398   \n",
       "met_patient5   3.101205 -0.693575  3.444901 -0.966174  1.443889 -0.173731   \n",
       "met_patient6   2.855980  0.787723  3.565637  0.719679  2.169149  2.559360   \n",
       "met_patient7  -0.035269  2.674077  1.923787  3.127244 -1.235824  2.743984   \n",
       "met_patient8   3.153689  1.715438  3.386103  2.503668  2.654869 -0.749824   \n",
       "met_patient9  -1.369572  3.800762 -0.177309  3.007730  2.531929  2.315384   \n",
       "met_patient10  1.997679  3.144291  2.932373  0.647715  2.305203 -1.472370   \n",
       "met_patient11  1.209936  3.451827  1.375446  2.356983 -1.411168  2.643656   \n",
       "met_patient12  3.481066 -0.289544  2.793193  0.248022  0.000000  2.183177   \n",
       "met_patient13  3.719666  1.002157  3.703100 -0.288530 -1.686133  2.459486   \n",
       "met_patient14  1.028050  0.479186  2.752916  0.240250  0.000000  0.530136   \n",
       "met_patient15  1.308003  0.215532  0.778021  0.443857  0.000000 -0.461929   \n",
       "met_patient16  2.717411  1.712626  3.099088  1.483767  0.000000  1.303103   \n",
       "met_patient17 -0.963371  1.496707  0.348947  2.091970 -1.294992  2.585932   \n",
       "met_patient18  1.741497  1.882959  2.416238  2.547161  0.000000 -0.174769   \n",
       "met_patient19  2.776374 -0.675923  2.472819 -0.726767 -0.367948 -0.876148   \n",
       "met_patient20 -1.563837  1.556350 -1.714443  2.449821  0.000000  0.832605   \n",
       "\n",
       "gene_name        HMGCS2     KRT13      DSG3     KRT14  ...    SEMA6C  \\\n",
       "0                                                      ...             \n",
       "met_patient1   1.186009  3.853465  1.490822  1.525784  ...  1.059692   \n",
       "met_patient2  -0.863597  0.574922  1.556667  3.864502  ...  0.627601   \n",
       "met_patient3   0.087568  1.213881  2.546487  4.447619  ...  0.606231   \n",
       "met_patient4   2.709300  3.093846  0.222066 -0.467246  ...  0.349627   \n",
       "met_patient5   0.415991  1.454945  1.997646  3.041708  ...  0.865187   \n",
       "met_patient6   0.508664  4.163777  2.718144  3.064817  ...  0.912817   \n",
       "met_patient7   2.487469  2.634731 -1.406714  1.325243  ...  0.487873   \n",
       "met_patient8   0.985624  3.207443  1.572401  2.394732  ...  1.181726   \n",
       "met_patient9   1.230280  3.240573 -1.950782  1.116309  ...  0.708106   \n",
       "met_patient10  2.299851  3.578766  0.059185  0.714925  ...  0.558036   \n",
       "met_patient11  2.336838  0.295193 -0.226360  1.575554  ...  0.711723   \n",
       "met_patient12 -0.833570  1.359475  1.699674  3.082666  ...  0.925812   \n",
       "met_patient13  1.813181  3.418172  2.660139  2.928010  ...  0.154424   \n",
       "met_patient14 -0.108965  3.454592 -0.680478 -0.018272  ...  0.858814   \n",
       "met_patient15 -0.934794  0.454555 -0.187020  2.129858  ...  0.374565   \n",
       "met_patient16 -0.348334  1.127898  1.216688  3.046589  ...  0.745535   \n",
       "met_patient17  2.474985  3.059174 -1.642065  1.369774  ...  1.007291   \n",
       "met_patient18  2.836458  3.425118  0.167288  1.778171  ...  0.586272   \n",
       "met_patient19 -0.377061 -0.131297 -0.142426  2.276137  ...  0.882120   \n",
       "met_patient20  1.841840  1.723930 -1.844664 -0.108686  ...  0.982416   \n",
       "\n",
       "gene_name        TRIM14    RNF19B      G6PD     SPART       CAT    TMEM65  \\\n",
       "0                                                                           \n",
       "met_patient1   1.523922  1.762159  1.778139  1.049319  1.775893  0.994396   \n",
       "met_patient2   1.249193  1.821425  1.486739  1.187772  2.167517  1.162418   \n",
       "met_patient3   1.454688  1.798896  2.091919  0.926867  1.780687  1.267814   \n",
       "met_patient4   1.470653  1.794031  2.121171  1.503799  2.153338  0.531185   \n",
       "met_patient5   1.732806  2.035776  1.774489  1.292398  1.829720  1.389975   \n",
       "met_patient6   1.562614  1.905865  1.897888  1.176277  1.564493  1.159456   \n",
       "met_patient7   1.488315  1.648729  1.913074  1.545804  2.150972  0.809176   \n",
       "met_patient8   1.607846  1.720087  1.964063  1.360141  1.793545  1.140722   \n",
       "met_patient9   1.004085  1.750707  1.518511  1.423754  1.696444  0.753154   \n",
       "met_patient10  1.388144  1.522396  1.687928  1.363710  2.130172  0.939829   \n",
       "met_patient11  0.953112  1.381830  1.603746  1.226966  1.742630  0.627438   \n",
       "met_patient12  1.394826  1.848327  1.734852  1.507281  1.594602  1.027366   \n",
       "met_patient13  1.684059  2.136211  1.744812  1.484694  2.245799  1.018368   \n",
       "met_patient14  1.345005  1.412870  1.832265  1.540573  1.591973  1.055818   \n",
       "met_patient15  1.567450  1.921197  1.875732  1.510821  1.487206  1.022486   \n",
       "met_patient16  1.808487  1.798279  2.201339  1.533009  1.533213  0.952279   \n",
       "met_patient17  1.504158  1.926353  1.530493  1.439683  1.947224  1.007167   \n",
       "met_patient18  1.058915  1.118536  1.689522  0.779510  1.662170  0.004579   \n",
       "met_patient19  1.330328  1.785747  2.204198  0.375206  1.411414  0.711875   \n",
       "met_patient20  1.693635  2.099919  1.756516  1.673222  2.366270  0.552778   \n",
       "\n",
       "gene_name      TNFRSF10A     TMTC2  metastasis  \n",
       "0                                               \n",
       "met_patient1    1.488546  1.190326           1  \n",
       "met_patient2    1.298386  0.722782           1  \n",
       "met_patient3    1.325008  0.331529           1  \n",
       "met_patient4    0.917736  0.722206           1  \n",
       "met_patient5    0.898215  0.634024           1  \n",
       "met_patient6    1.589704  1.295030           1  \n",
       "met_patient7    1.157566  1.140637           1  \n",
       "met_patient8    1.396809  1.027810           1  \n",
       "met_patient9    0.733927  1.140524           1  \n",
       "met_patient10   1.181572  1.117921           1  \n",
       "met_patient11   0.698353  1.039676           1  \n",
       "met_patient12   0.794174  0.867019           1  \n",
       "met_patient13   1.547062  0.655667           1  \n",
       "met_patient14   0.898572  1.056024           1  \n",
       "met_patient15   1.020046  0.553325           1  \n",
       "met_patient16   1.465845  0.472917           1  \n",
       "met_patient17   0.813094  1.593850           1  \n",
       "met_patient18   0.831678  0.634356           1  \n",
       "met_patient19   0.668265  0.459408           1  \n",
       "met_patient20   0.973437  1.328308           1  \n",
       "\n",
       "[20 rows x 10001 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a \"metastasis\" label column for each category (metastasis / no metastasis)\n",
    "prep_df.loc[met_labels, \"metastasis\"] = 1\n",
    "prep_df.loc[no_met_labels, \"metastasis\"] = 0\n",
    "# Convert \"metastasis\" column to type int\n",
    "prep_df[\"metastasis\"] = prep_df[\"metastasis\"].astype(\"int\")\n",
    "prep_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c6c5b614",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(160, 10000)\n",
      "(160,)\n"
     ]
    }
   ],
   "source": [
    "# Seperate independent variables\n",
    "X = prep_df.drop(labels = [\"metastasis\"], axis = 1)\n",
    "print(X.shape)\n",
    "# Extract dependent variable\n",
    "Y = prep_df[\"metastasis\"]\n",
    "print(Y.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3423fad4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>frequency_in_top_100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FKBP6</td>\n",
       "      <td>430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>ASIC5</td>\n",
       "      <td>382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>MAPK8IP1</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>F11R</td>\n",
       "      <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>PABPC5</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>SLC5A1</td>\n",
       "      <td>288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>CCDC33</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>ONECUT1</td>\n",
       "      <td>260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>CLC</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>ZNF467</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>KCNS3</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>CDH2</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>PTRH1</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>749</th>\n",
       "      <td>NKAIN3</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>CGREF1</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OLFM4</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>ATP4A</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>PCDHB7</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>CSPG5</td>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SHROOM3</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>MND1</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>KCNQ4</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>ELF5</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>ZNF285</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>GPR35</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>SPATA13</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>SPTA1</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>FADS6</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>FOXA2</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>NGFR</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>EPOR</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>VWA5A</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HNRNPCL1</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>705</th>\n",
       "      <td>KCNE1</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>FAM241B</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>IL18</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830</th>\n",
       "      <td>DTNA</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1006</th>\n",
       "      <td>SLC24A3</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>SPTBN5</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>TAS1R3</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>OPCML</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CARD9</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>RHOF</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>621</th>\n",
       "      <td>PDLIM4</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>PDE8B</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>562</th>\n",
       "      <td>CRYM</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>SMIM41</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>VPS37D</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NINL</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>MTRNR2L10</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name  frequency_in_top_100\n",
       "0         FKBP6                   430\n",
       "313       ASIC5                   382\n",
       "138    MAPK8IP1                   332\n",
       "280        F11R                   310\n",
       "449      PABPC5                   297\n",
       "53       SLC5A1                   288\n",
       "24       CCDC33                   273\n",
       "534     ONECUT1                   260\n",
       "100         CLC                   258\n",
       "125      ZNF467                   256\n",
       "740       KCNS3                   253\n",
       "611        CDH2                   253\n",
       "1086      PTRH1                   239\n",
       "749      NKAIN3                   232\n",
       "596      CGREF1                   224\n",
       "3         OLFM4                   219\n",
       "310       ATP4A                   211\n",
       "1117     PCDHB7                   206\n",
       "39        CSPG5                   201\n",
       "5       SHROOM3                   192\n",
       "160        MND1                   192\n",
       "12        KCNQ4                   192\n",
       "955        ELF5                   187\n",
       "348      ZNF285                   185\n",
       "109       GPR35                   182\n",
       "41      SPATA13                   181\n",
       "116       SPTA1                   174\n",
       "906       FADS6                   172\n",
       "447       FOXA2                   172\n",
       "106        NGFR                   171\n",
       "461        EPOR                   170\n",
       "537       VWA5A                   169\n",
       "1      HNRNPCL1                   169\n",
       "705       KCNE1                   165\n",
       "291     FAM241B                   165\n",
       "286        IL18                   161\n",
       "830        DTNA                   153\n",
       "1006    SLC24A3                   151\n",
       "132      SPTBN5                   151\n",
       "422      TAS1R3                   150\n",
       "516       OPCML                   149\n",
       "23        CARD9                   148\n",
       "130        RHOF                   148\n",
       "621      PDLIM4                   147\n",
       "118       PDE8B                   147\n",
       "562        CRYM                   146\n",
       "80       SMIM41                   144\n",
       "142      VPS37D                   143\n",
       "22         NINL                   141\n",
       "536   MTRNR2L10                   141"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Gene ranking using Random Forest Classifier (Metastasis vs. No-Metastasis)\n",
    "genes_selection_df = rf_model(X,Y)\n",
    "genes_selection_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5f3dfc85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>frequency_in_top_100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FKBP6</td>\n",
       "      <td>430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASIC5</td>\n",
       "      <td>382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAPK8IP1</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F11R</td>\n",
       "      <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PABPC5</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SLC5A1</td>\n",
       "      <td>288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CCDC33</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ONECUT1</td>\n",
       "      <td>260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CLC</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ZNF467</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>KCNS3</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDH2</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PTRH1</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NKAIN3</td>\n",
       "      <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CGREF1</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>OLFM4</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ATP4A</td>\n",
       "      <td>211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PCDHB7</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CSPG5</td>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SHROOM3</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MND1</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>KCNQ4</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ELF5</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ZNF285</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GPR35</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>SPATA13</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SPTA1</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>FADS6</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FOXA2</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NGFR</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>EPOR</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>VWA5A</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HNRNPCL1</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>KCNE1</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>FAM241B</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>IL18</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>DTNA</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SLC24A3</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SPTBN5</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>TAS1R3</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>OPCML</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CARD9</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>RHOF</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>PDLIM4</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>PDE8B</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>CRYM</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>SMIM41</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>VPS37D</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NINL</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>MTRNR2L10</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name  frequency_in_top_100\n",
       "0       FKBP6                   430\n",
       "1       ASIC5                   382\n",
       "2    MAPK8IP1                   332\n",
       "3        F11R                   310\n",
       "4      PABPC5                   297\n",
       "5      SLC5A1                   288\n",
       "6      CCDC33                   273\n",
       "7     ONECUT1                   260\n",
       "8         CLC                   258\n",
       "9      ZNF467                   256\n",
       "10      KCNS3                   253\n",
       "11       CDH2                   253\n",
       "12      PTRH1                   239\n",
       "13     NKAIN3                   232\n",
       "14     CGREF1                   224\n",
       "15      OLFM4                   219\n",
       "16      ATP4A                   211\n",
       "17     PCDHB7                   206\n",
       "18      CSPG5                   201\n",
       "19    SHROOM3                   192\n",
       "20       MND1                   192\n",
       "21      KCNQ4                   192\n",
       "22       ELF5                   187\n",
       "23     ZNF285                   185\n",
       "24      GPR35                   182\n",
       "25    SPATA13                   181\n",
       "26      SPTA1                   174\n",
       "27      FADS6                   172\n",
       "28      FOXA2                   172\n",
       "29       NGFR                   171\n",
       "30       EPOR                   170\n",
       "31      VWA5A                   169\n",
       "32   HNRNPCL1                   169\n",
       "33      KCNE1                   165\n",
       "34    FAM241B                   165\n",
       "35       IL18                   161\n",
       "36       DTNA                   153\n",
       "37    SLC24A3                   151\n",
       "38     SPTBN5                   151\n",
       "39     TAS1R3                   150\n",
       "40      OPCML                   149\n",
       "41      CARD9                   148\n",
       "42       RHOF                   148\n",
       "43     PDLIM4                   147\n",
       "44      PDE8B                   147\n",
       "45       CRYM                   146\n",
       "46     SMIM41                   144\n",
       "47     VPS37D                   143\n",
       "48       NINL                   141\n",
       "49  MTRNR2L10                   141"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_selection_df = genes_selection_df.reset_index(drop=True)\n",
    "genes_selection_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e9ed5b80",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_selection_df.to_csv(\"top_genes_BLCA_tpm.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b8bfacc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65ec247e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
